source/common/upstream/cluster_manager_impl.cc

```diff
@@ -343,6 +353,80 @@ void ClusterManagerImpl::onClusterInit(Cluster& cluster) {
   }
 }
 
+bool ClusterManagerImpl::scheduleUpdate(const Cluster& cluster, uint32_t priority,
+                                        const HostVector& hosts_added,
+                                        const HostVector& hosts_removed) {
+  const auto& update_merge_window = cluster.info()->lbConfig().update_merge_window();
+  const auto timeout = DurationUtil::durationToMilliseconds(update_merge_window);
+  PendingUpdatesByPriorityMapPtr updates_by_prio;
+  PendingUpdatesPtr updates;
+
+  // Find pending updates for this cluster.
+  auto updates_by_prio_it = updates_map_.find(cluster.info()->name());
+  if (updates_by_prio_it != updates_map_.end()) {
+    updates_by_prio = updates_by_prio_it->second;
+  } else {
+    updates_by_prio = std::make_shared<PendingUpdatesByPriorityMap>();
+    updates_map_[cluster.info()->name()] = updates_by_prio;
+  }
+
+  // Find pending updates for this priority.
+  auto updates_it = updates_by_prio->find(priority);
+  if (updates_it != updates_by_prio->end()) {
+    updates = updates_it->second;
+  } else {
+    updates = std::make_shared<PendingUpdates>();
+    (*updates_by_prio)[priority] = updates;
+  }
+
+  // Has an update_merge_window gone by since the last update? If so, don't schedule
+  // the update so it can be applied immediately.
+  auto delta = std::chrono::steady_clock::now() - updates->last_updated;
+  uint64_t delta_ms = std::chrono::duration_cast<std::chrono::milliseconds>(delta).count();
+  if (delta_ms > timeout) {
+    updates->last_updated = std::chrono::steady_clock::now();
+    return false;
+  }
+
+  // Record the updates — preserving order — that should be applied when the timer fires.
```

Comment:
FWIW, I was arguing for just a deterministic ordering, e.g. using `set` vs. `unordered_set` in my earlier comment, rather than arrival ordering. But, I don't mind if we stick with this.